doc/sphinxext/custom_references_resolver.py

```diff
@@ -0,0 +1,83 @@
+"""Adapted from
+sphinx.transforms.post_transforms.ReferencesResolver.resolve_anyref
+
+If 'py' is one of the domains and `py:class` is defined,
+the Python domain will be processed before the 'std' domain.
+"""
+from contextlib import suppress
+
+from docutils import nodes
+from sphinx.transforms.post_transforms import ReferencesResolver
+
+
+class CustomReferencesResolver(ReferencesResolver):
+    def resolve_anyref(self, refdoc, node, contnode):
+        """Resolve reference generated by the "any" role."""
+        stddomain = self.env.get_domain('std')
+        target = node['reftarget']
+
+        # first, try resolving as :doc:
+        doc_ref = stddomain.resolve_xref(self.env, refdoc, self.app.builder,
+                                         'doc', target, node, contnode)
+        if doc_ref:
+            return self.create_node(('doc', doc_ref))
+
+        # process 'py' domain first for python classes
+        if "py:class" in node:
+            with suppress(KeyError):
+                py_domain = self.env.domains['py']
+                py_ref = py_domain.resolve_any_xref(
+                    self.env, refdoc, self.app.builder, target, node, contnode)
+                if py_ref:
+                    return self.create_node(py_ref[0])
+
+        # next, do the standard domain (makes this a priority)
+        std_ref = stddomain.resolve_any_xref(
+            self.env, refdoc, self.app.builder, target, node, contnode)
+        if std_ref:
+            return self.create_node(std_ref[0])
+
+        for domain in self.env.domains.values():
+            try:
+                ref = domain.resolve_any_xref(
+                    self.env, refdoc, self.app.builder, target, node, contnode)
+                if ref:
+                    return self.create_node(ref[0])
+            except NotImplementedError:
+                # the domain doesn't yet support the new interface
+                # we have to manually collect possible references (SLOW)
+                for role in domain.roles:
+                    res = domain.resolve_xref(self.env, refdoc,
+                                              self.app.builder, role, target,
+                                              node, contnode)
+                    if res and isinstance(res[0], nodes.Element):
+                        result = ('%s:%s' % (domain.name, role), res)
+                        return self.create_node(result)
+        # no results
+        return None
+
+    def create_node(self, result):
+        res_role, newnode = result
+        # Override "any" class with the actual role type to get the styling
```

Comment:
Nice!! One confusing thing is that `:term:` does not generally require monospace, but some terms are explicitly formatted with double backticks in the glossary (but the link target does not include backticks). I assume we're not really able to make that work nicely. No big deal.

I can't see what code controls styling when there's no match.

Reply:
When the reference is missing the `missing-reference` event is called. This is handled in another PR: https://github.com/scikit-learn/scikit-learn/pull/12970/files. In that PR, the single backticks are treated as double backticks when there are not matches.